src/Npgsql/NpgsqlConnection.cs

```diff
@@ -198,21 +194,23 @@ void GetPoolAndSettings()
                 }
 
                 // Maybe pooling is off
-                if (!Settings.Pooling)
-                    return;
-
-                // Connstring may be equivalent to one that has already been seen though (e.g. different
-                // ordering). Have NpgsqlConnectionStringBuilder produce a canonical string representation
-                // and recheck.
-                var canonical = Settings.ConnectionString;
-                if (pools.TryGetValue(canonical, out _pool))
-                    pools[_connectionString] = _pool;
-                else
+                if (Settings.Pooling)
                 {
-                    // Really unseen, need to create a new pool
-                    _pool = pools[_connectionString] = new ConnectorPool(Settings, canonical);
-                    if (_connectionString != canonical)
-                        pools[canonical] = _pool;
+                    // Connstring may be equivalent to one that has already been seen though (e.g. different
+                    // ordering). Have NpgsqlConnectionStringBuilder produce a canonical string representation
+                    // and recheck.
+                    var canonical = Settings.ConnectionString;
+
+                    if (pools.TryGetValue(canonical, out _pool))
+                        _pool = pools.GetOrAdd(_connectionString, _pool); // Assign the connection string the canonical pool. If someone beat us to it use whatever is there already.
```

Comment:
And one last time, if the canonical pool is found, simply return from the method after setting the non-canonical key.